picoquic/bytestream.h

```diff
@@ -0,0 +1,77 @@
+/*
+* Author: Christian Huitema
+* Copyright (c) 2018, Private Octopus, Inc.
+* All rights reserved.
+*
+* Permission to use, copy, modify, and distribute this software for any
+* purpose with or without fee is hereby granted, provided that the above
+* copyright notice and this permission notice appear in all copies.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL Private Octopus, Inc. BE LIABLE FOR ANY
+* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef PICOQUIC_BYTESTREAM_H
+#define PICOQUIC_BYTESTREAM_H
+
+#include "picoquic_internal.h"
+
+#define BYTESTREAM_MAX_BUFFER_SIZE 2048
```

Comment:
It would be nice to add comments about the principles of operation. I understand that you want to do buffered IO without using FILE from stdio, but when I read the code that's not clear. Do you intend to have one buffer per frame, or per file? In the read case, how do you manage overlapping buffers?

Reply:
Have a look in picoquictest/bytestream_test.c on how to use the bytestream.

The connection to the file IO is really only that the resulting memory buffers can be read or written to an from files.

I'll open another PR on how to use the bytestream for binary file logging so you get an idea how that is intended to work.